File: design-patterns/template-method/ruby/html_report.rb
require_relative 'report'

class HTMLReport < Report
  def output_start
    p '<html>'
  end
  
  def output_head
    p '<head>'
    p "<title>#{title}</title>"
    p '</head>'
  end
  
  def output_body_start
    p '<body>'
  end
  
  def output_line(line)
    p "<p>#{line}</p>"
  end
  
  def output_body_end
    p '</body>'
  end
  
  def output_end
    p '</html>'
  end
end

File: design-patterns/cs/Patrones/Patrones/TemplateMethod/HTMLReport.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace Patrones.TemplateMethod
{
    public class HTMLReport : Report
    {
        protected override string OutputStart()
        {
            return "<html>";
        }

        protected override string OutputHead()
        {
            string aux = string.Empty;

            aux += "<head>";
            aux += "<title>";
            aux += Title;
            aux += "</title>";
            aux += "</head>";

            return aux;
        }

        protected override string OutputBodyStart()
        {
            return "<body>";
        }

        protected override string OutputBodyEnd()
        {
           return  "</body>";
        }

        protected override string OutputEnd()
        {
            return "</html>";
        }

        protected override string OutputLine(string line)
        {
            return "<p>" + line + "</p>";
        }
    }
}
